clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java

```diff
@@ -110,7 +116,68 @@ public String toString() {
         }
     }
 
-    private Optional<Integer> optionalEpoch(int rawEpochValue) {
+    public static final class UnresolvedPartitions {
```

Comment:
The whole FetchRequest class is quite hard to follow without reading the KIP and looking at multiple places. It will be good to add some comments at the class level.